daemon/logger/awslogs/cloudwatchlogs.go

```diff
@@ -83,18 +120,65 @@ type wrappedEvent struct {
 	inputLogEvent *cloudwatchlogs.InputLogEvent
 	insertOrder   int
 }
-type byTimestamp []wrappedEvent
 
-// init registers the awslogs driver
-func init() {
-	if err := logger.RegisterLogDriver(name, New); err != nil {
-		logrus.Fatal(err)
-	}
-	if err := logger.RegisterLogOptValidator(name, ValidateLogOpt); err != nil {
-		logrus.Fatal(err)
+// eventBatch holds the events that are batched for submission and the
+// associated data about it.  This type is not threadsafe and must not
+// be used concurrently.
+type eventBatch struct {
+	events []wrappedEvent
+	bytes  int
+}
+
+func newEventBatch() *eventBatch {
+	return &eventBatch{
+		events: make([]wrappedEvent, 0),
+		bytes:  0,
 	}
 }
 
+// Events returns the sorted slice of wrappedEvents.
+func (b *eventBatch) Events() []wrappedEvent {
+	// events in a batch must be sorted by timestamp
+	// see http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
+	sort.Sort(byTimestamp(b.events))
+	return b.events
+}
+
+// Add adds an event to the batch of events. This adds in the
+// necessary overhead for an event to be logged.
+func (b *eventBatch) Add(event wrappedEvent, size int) {
+	b.bytes = b.bytes + size + perEventBytes
+	b.events = append(b.events, event)
```

Comment:
Looks unsafe.

Reply:
Same case with `Count` and `Size`.

Reply:
Indeed. This type isn't intended for concurrently use and isn't accessed concurrently as it stands - hence the unsafe operations. Do you have suggestions to improve this given the current implementation of its consumer? My goal here was to keep the type small and simple for this use case, but I am happy to add some mechanisms for safety.

Reply:
May need a mention in the function description; also given that it's only use in this file, we should probably un-export these methods

Reply:
Sounds good. This type is only used here so I'll unexport the methods. I'll also include a statement about the thread unsafe nature in each method's GoDoc line to make sure future developers understand this design assumption.